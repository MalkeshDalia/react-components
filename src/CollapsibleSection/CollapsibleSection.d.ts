import { CSSProperties, PureComponent } from 'react';

export type CollapsibleSectionProps = {
  children: any, // TODO: t.ReactChildren
  isOpen: boolean,
  isSelected?: boolean,
  onChange: (isOpen: boolean) => void,
  onOpen?: () => void,
  onClose?: () => void,
  header?: any, // TODO: t.ReactChild
  icons?: {
    open?: string,
    closed?: string
  },
  className?: string,
  id?: string,
  style?: CSSProperties
}

export default class CollapsibleSection extends PureComponent<CollapsibleSectionProps> {}
