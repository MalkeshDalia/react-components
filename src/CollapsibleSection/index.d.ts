import CollapsibleSection, { CollapsibleSectionProps } from './CollapsibleSection';
export default CollapsibleSection;
export { CollapsibleSectionProps };
