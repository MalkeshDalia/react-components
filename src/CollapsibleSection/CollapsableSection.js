import React from 'react';
import cx from 'classnames';
import { skinnable, props, t } from '../utils';
import FlexView from 'react-flexview';
import Icon from '../Icon/Icon';

export const Props = {
  children: t.ReactChildren,
  isOpen: t.Boolean,
  isSelected: t.maybe(t.Boolean),
  onChange: t.Function,
  onOpen: t.maybe(t.Function),
  onClose: t.maybe(t.Function),
  header: t.maybe(t.ReactChild),
  icons: t.maybe(t.struct({
    open: t.maybe(t.String),
    closed: t.maybe(t.String)
  })),
  className: t.maybe(t.String),
  id: t.maybe(t.String),
  style: t.maybe(t.Object)
};

/** A collapsable panel, allowing you to toggle more/less content
 * @param children - panel content (visible only when expanded)
 * @param isOpen - true if panel is expanded
 * @param isSelected - true if panel has focus
 * @param onChange - called when panel is toggled
 * @param onOpen - called when panel is expanded
 * @param onClose - called when panel is collapsed
 * @param header - header content (the only visible part when panel is collapsed)
 * @param icons - icons for open/closed panel
 */
@skinnable()
@props(Props)
export default class CollapsableSection extends React.Component {

  static defaultProps = {
    onOpen: () => {},
    onClose: () => {}
  };

  _onChange = () => {
    const { onChange, onOpen, onClose, isOpen } = this.props;
    onChange(!isOpen);
    if (isOpen) {
      onClose();
    } else {
      onOpen();
    }
  };

  getLocals() {
    const {
      props: {
        isOpen, isSelected,
        children, header, icons,
        className, id, style
      },
      _onChange: onChange
    } = this;

    return {
      isOpen,
      children,
      header,
      icons,
      onChange,
      wrapperProps: {
        id,
        style,
        className: cx(className, 'collapsable-section', { 'is-open': isOpen, 'is-selected': isSelected })
      }
    };
  }

  templateIcons = ({ isOpen, icons }) => (
    icons && <Icon icon={isOpen ? icons.open : icons.closed} />
  );

  template({ isOpen, children, onChange, header, icons, wrapperProps }) {
    return (
      <FlexView {...wrapperProps} column>
        <FlexView className='content' vAlignContent='center' onClick={onChange}>
          {header}
          <FlexView marginLeft='auto' shrink={false}>
            {this.templateIcons({ isOpen, icons })}
          </FlexView>
        </FlexView>
        {isOpen && children}
      </FlexView>
    );
  }

}
