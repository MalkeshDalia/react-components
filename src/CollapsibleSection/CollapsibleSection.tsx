import * as React from 'react';
import cx from '../utils/classnames';
import { props, t, ReactChildren, ReactChild } from '../utils';
import FlexView from 'react-flexview';
import Icon from '../Icon/Icon';

export type CollapsibleSectionRequiredProps = {
  /** panel content (visible only when expanded) */
  children: JSX.Element,
  /** true if panel is expanded */
  isOpen: boolean,
  /** true if panel has focus */
  isSelected?: boolean,
  /** called when panel is toggled */
  onChange: (isOpen: boolean) => void,
  /** header content (the only visible part when panel is collapsed) */
  header?: JSX.Element,
  /** icons for open/closed panel */
  icons?: {
    open?: string,
    closed?: string
  },
  /** an optional class name to pass to top level element of the component */
  className?: string,
  /** an optional style object to pass to top level element of the component */
  style?: React.CSSProperties,
  id?: string,
};

export type CollapsibleSectionDefaultProps = {
  /** called when panel is expanded */
  onOpen: () => void,
  /** called when panel is collapsed */
  onClose: () => void,
};

export type CollapsibleSectionProps = CollapsibleSectionRequiredProps & Partial<CollapsibleSectionDefaultProps>;
type CollapsibleSectionDefaultedProps = CollapsibleSectionRequiredProps & CollapsibleSectionDefaultProps;

export const Props = {
  children: ReactChildren,
  isOpen: t.Boolean,
  isSelected: t.maybe(t.Boolean),
  onChange: t.Function,
  onOpen: t.maybe(t.Function),
  onClose: t.maybe(t.Function),
  header: t.maybe(ReactChild),
  icons: t.maybe(t.struct({
    open: t.maybe(t.String),
    closed: t.maybe(t.String)
  })),
  className: t.maybe(t.String),
  style: t.maybe(t.Object),
  id: t.maybe(t.String)
};

/** A collapsible panel, allowing you to toggle more/less content */
@props(Props)
export default class CollapsibleSection extends React.PureComponent<CollapsibleSectionProps> {

  static defaultProps: CollapsibleSectionDefaultProps = {
    onOpen: () => {},
    onClose: () => {}
  };

  onChange = () => {
    const { onChange, onOpen, onClose, isOpen } = this.props as CollapsibleSectionDefaultedProps;
    onChange(!isOpen);
    if (isOpen) {
      onClose();
    } else {
      onOpen();
    }
  };

  render() {
    const {
      isOpen, isSelected,
      children, header, icons,
      className, id, style
    } = this.props as CollapsibleSectionDefaultedProps;
    const { onChange } = this;

    const wrapperProps = {
      id,
      style,
      className: cx(className, 'collapsible-section', { 'is-open': isOpen, 'is-selected': isSelected })
    };

    return (
      <FlexView {...wrapperProps} column>
        <FlexView className='content' vAlignContent='center' onClick={onChange}>
          {header}
          <FlexView marginLeft='auto' shrink={false}>
            {icons && <Icon icon={isOpen ? icons.open : icons.closed} />}
          </FlexView>
        </FlexView>
        {isOpen && children}
      </FlexView>
    );
  }

}
