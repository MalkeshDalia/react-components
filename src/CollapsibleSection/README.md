# CollapsableSection

A collapsable panel, allowing you to toggle more/less content

## Props
|Name|Type|Default|Description|
|----|----|-------|-----------|
| **children** | <code>ReactChildren</code> |  | **required**. Panel content (visible only when expanded) |
| **isOpen** | <code>Boolean</code> |  | **required**. True if panel is expanded |
| **isSelected** | <code>Boolean</code> |  | *optional*. True if panel has focus |
| **onChange** | <code>Function</code> |  | **required**. Called when panel is toggled |
| **onOpen** | <code>Function</code> | <code>"onOpen"</code> | *optional*. Called when panel is expanded |
| **onClose** | <code>Function</code> | <code>"onClose"</code> | *optional*. Called when panel is collapsed |
| **header** | <code>ReactChild</code> |  | *optional*. Header content (the only visible part when panel is collapsed) |
| **icons** | <code>Struct{open: ?String, closed: ?String}</code> |  | *optional*. Icons for open/closed panel |
| **className** | <code>String</code> |  | *optional*. Additional `className` for wrapper element |
| **id** | <code>String</code> |  | *optional*. Custom `id` for wrapper element |
| **style** | <code>Object</code> |  | *optional*. Inline-style overrides for wrapper element |