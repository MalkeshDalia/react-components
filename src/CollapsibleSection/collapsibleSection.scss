@import '../utils/theme/variables.scss';
$background: transparent !default;
$header-background: $brc-gradientGrey !default;
$header-background-open: $brc-gradientGrey !default;
$header-background-selected: transparent !default;
$header-background-hover: $brc-gradientDarkGrey !default;
$border-color: $brc-silver !default;

.collapsible-section {
  border: 1px solid $border-color;
  background: $background;
  cursor: pointer;

  .content {
    padding: 0 19px;
    height: 100%;
    background: $header-background;

    &:hover {
      background: $header-background-hover;
    }
  }

  &.is-open {
    .content {
      border-bottom: 1px solid $border-color;
      background: $header-background-open;

      &:hover {
        background: $header-background-hover;
      }
    }
  }

  &.is-selected,
  &.is-open.is-selected {
    background: $header-background-selected;

    &:hover {
      background: $header-background-hover;
    }
  }
}
