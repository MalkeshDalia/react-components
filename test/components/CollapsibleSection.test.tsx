import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import CollapsibleSection from '../../src/CollapsibleSection';

let consoleError: jest.SpyInstance<{}>;

beforeAll(() => {
  consoleError = jest.spyOn(console, 'error');
});

afterEach(() => {
  expect(consoleError).not.toHaveBeenCalled();
});

const content = <div>HI</div>;

describe('CollapsibleSection', () => {

  it('renders correctly when open', () => {
    const component = renderer.create(
      <CollapsibleSection isOpen={true} onChange={() => {}}>
        {content}
      </CollapsibleSection>
    );
    expect(component).toMatchSnapshot();
  });

  it('renders correctly when closed', () => {
    const component = renderer.create(
      <CollapsibleSection isOpen={false} onChange={() => {}}>
        {content}
      </CollapsibleSection>
    );
    expect(component).toMatchSnapshot();
  });

  it('computes className when open', () => {
    const component = shallow(
      <CollapsibleSection className='awesome-class' isOpen={true} onChange={() => {}}>
        {content}
      </CollapsibleSection>
    );
    expect(component.hasClass('collapsible-section')).toBe(true);
    expect(component.hasClass('awesome-class')).toBe(true);
    expect(component.hasClass('is-open')).toBe(true);
  });

  it('computes className when closed', () => {
    const component = shallow(
      <CollapsibleSection className='awesome-class' isOpen={false} onChange={() => {}}>
        {content}
      </CollapsibleSection>
    );
    expect(component.hasClass('collapsible-section')).toBe(true);
    expect(component.hasClass('awesome-class')).toBe(true);
    expect(!component.hasClass('is-open')).toBe(true);
  });

  it('invokes onOpen and onChange when opened', () => {
    const onOpen = jest.fn();
    const onClose = jest.fn();
    const onChange = jest.fn();
    const component = shallow(
      <CollapsibleSection className='awesome-class' isOpen={false} onChange={onChange} onOpen={onOpen} onClose={onClose}>
        {content}
      </CollapsibleSection>
    );
    component.find('.content').simulate('click');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toBe(true);
    expect(onOpen).toHaveBeenCalledTimes(1);
    expect(onOpen).toHaveBeenCalledTimes(1);
    expect(onClose).not.toHaveBeenCalled();
  });

  it('invokes onClose and onChange when closed', () => {
    const onOpen = jest.fn();
    const onClose = jest.fn();
    const onChange = jest.fn();
    const component = shallow(
      <CollapsibleSection className='awesome-class' isOpen={true} onChange={onChange} onOpen={onOpen} onClose={onClose}>
        {content}
      </CollapsibleSection>
    );
    component.find('.content').simulate('click');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toBe(false);
    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onClose).toHaveBeenCalledTimes(1);
    expect(onOpen).not.toHaveBeenCalled();
  });

});
