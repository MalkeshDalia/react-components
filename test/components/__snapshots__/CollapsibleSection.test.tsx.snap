// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleSection renders correctly when closed 1`] = `
<div
  className="react-flex-view flex-column collapsible-section"
  id={undefined}
  style={
    Object {
      "MozBoxFlex": "0 1 auto",
      "WebkitBoxFlex": "0 1 auto",
      "WebkitFlex": "0 1 auto",
      "flex": "0 1 auto",
      "msFlex": "0 1 auto",
    }
  }
>
  <div
    className="react-flex-view align-content-center content"
    onClick={[Function]}
    style={
      Object {
        "MozBoxFlex": "0 1 auto",
        "WebkitBoxFlex": "0 1 auto",
        "WebkitFlex": "0 1 auto",
        "flex": "0 1 auto",
        "msFlex": "0 1 auto",
      }
    }
  >
    <div
      className="react-flex-view"
      style={
        Object {
          "MozBoxFlex": "0 0 auto",
          "WebkitBoxFlex": "0 0 auto",
          "WebkitFlex": "0 0 auto",
          "flex": "0 0 auto",
          "marginLeft": "auto",
          "msFlex": "0 0 auto",
        }
      }
    />
  </div>
</div>
`;

exports[`CollapsibleSection renders correctly when open 1`] = `
<div
  className="react-flex-view flex-column collapsible-section is-open"
  id={undefined}
  style={
    Object {
      "MozBoxFlex": "0 1 auto",
      "WebkitBoxFlex": "0 1 auto",
      "WebkitFlex": "0 1 auto",
      "flex": "0 1 auto",
      "msFlex": "0 1 auto",
    }
  }
>
  <div
    className="react-flex-view align-content-center content"
    onClick={[Function]}
    style={
      Object {
        "MozBoxFlex": "0 1 auto",
        "WebkitBoxFlex": "0 1 auto",
        "WebkitFlex": "0 1 auto",
        "flex": "0 1 auto",
        "msFlex": "0 1 auto",
      }
    }
  >
    <div
      className="react-flex-view"
      style={
        Object {
          "MozBoxFlex": "0 0 auto",
          "WebkitBoxFlex": "0 0 auto",
          "WebkitFlex": "0 0 auto",
          "flex": "0 0 auto",
          "marginLeft": "auto",
          "msFlex": "0 0 auto",
        }
      }
    />
  </div>
  <div>
    HI
  </div>
</div>
`;
